### PR TITLE
[NP-7477] Add RequirementsRuleAction

### PR DESCRIPTION
**Pull request by KernelDeimos**

### CODE REVIEW

File: src/foam/nanos/crunch/lite/RequirementsRuleAction.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+            for ( String id : getAvailable() ) {
+              var payload = new CapabilityJunctionPayload();
+              payload.setCapability(id);
+              capablePayloadDAO.put(payload);
+            }
+
+            capable.setCapabilityIds(getRequirements());
+          }
+        }, "");
```
Comment (ghost) [inline src/foam/nanos/crunch/lite/RequirementsRuleAction.js:77]: should probably describe the action (we do this elsewhere)

File: src/foam/nanos/crunch/lite/RequirementsRuleAction.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            var capable = (Capable) obj;
+            var capablePayloadDAO = capable.getCapablePayloadDAO(x);
+
+            for ( String id : getRequirements() ) {
+              var payload = new CapabilityJunctionPayload();
+              payload.setCapability(id);
+              capablePayloadDAO.put(payload);
+            }
+
+            for ( String id : getAvailable() ) {
```
Comment (ghost) [inline src/foam/nanos/crunch/lite/RequirementsRuleAction.js:71]: code in this for loop identical as above - comment please.

File: src/foam/nanos/crunch/lite/RequirementsRuleAction.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
+foam.CLASS({
```
Comment (kgrgreer) [inline src/foam/nanos/crunch/lite/RequirementsRuleAction.js:7]: + copyright

File: src/foam/nanos/crunch/lite/RequirementsRuleAction.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    'foam.nanos.ruler.RuleAction'
+  ],
+
+  javaImports: [
+    'foam.core.ContextAgent',
+    'foam.core.X',
+    'foam.dao.DAO',
+    'foam.nanos.crunch.CapabilityJunctionPayload',
+    'foam.nanos.alarming.Alarm',
+    'foam.nanos.alarming.AlarmReason',
```
Comment (kgrgreer) [inline src/foam/nanos/crunch/lite/RequirementsRuleAction.js]: no trailing comma

File: src/foam/nanos/crunch/lite/RequirementsRuleAction.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  ],
+
+  methods: [
+    {
+      name: 'applyAction',
+      javaCode: `
+        agency.submit(x, new ContextAgent() {
+          @Override
+          public void execute(X x) {
+            if ( ! (obj instanceof Capable) ) {
```
Comment (Nauna) [inline src/foam/nanos/crunch/lite/RequirementsRuleAction.js:48]: hmm - you are limited to Capable properties if you do this. You know you are getting a TPI obj. You should cast it accordingly.

File: src/foam/nanos/crunch/lite/RequirementsRuleAction.js (hunk 2 above)
Comment (Nauna) [reply src/foam/nanos/crunch/lite/RequirementsRuleAction.js:71]: what is available suppose to be?

File: src/foam/nanos/crunch/lite/RequirementsRuleAction.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+              capablePayloadDAO.put(payload);
+            }
+
+            for ( String id : getAvailable() ) {
+              var payload = new CapabilityJunctionPayload();
+              payload.setCapability(id);
+              capablePayloadDAO.put(payload);
+            }
+
+            capable.setCapabilityIds(getRequirements());
```
Comment (Nauna) [inline src/foam/nanos/crunch/lite/RequirementsRuleAction.js]: doing this will overide anything already set on CapabilityIds... consider putting into loop

**Review by Nauna — COMMENTED** [review]
-

File: src/foam/nanos/crunch/lite/RequirementsRuleAction.js (hunk 2 above)
Comment (KernelDeimos) [reply src/foam/nanos/crunch/lite/RequirementsRuleAction.js:71]: I'm not sure how to document it more clearly so I added a comment saying that there's documentation on the `available` property

**Pull request merged**